llext: Shell sample fixups

teburd opened the pull request:
Copies and modifies the hello_world sample extension for use with the
sample app.

Uses west+cmake+ninja in the steps to now show how extensions are built ensuring the sample doesn't bit rot quite as fast hopefully!

Fixes #79318

Commit:
llext: Hello world for the shell_loader sample

Hello world from the tests is not longer what it once was, a simple
hello_world function that prints. The sample was here to let people
explore and learn about how llext and ELFs work.

Moving the sample extension code to the shell loader sample disconnects
the two and lets the sample remain simpler.

Additionally we use west (cmake) and ninja directly using the zephyr
provided add_llext_target() to show how extensions can be built and
loaded. This is needed now as linker scripts and custom sections are
required for symbol exports.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

pillo79 commented:
Nice idea, appreciated! My 2c: I kind of regret pushing for creating the "simple" directory, since we have never found the need for a different test suite. Unless you already have ideas on that topic, we could remove the subfolder altogether and just have `tests/subsys/llext/` ...

markus-becker-tridonic-com commented:
@teburd I was trying this on an nRF52840DK (using Nordic Connect SDK and the bundled Zephyr) and cannot get it to work. 

I have added `CONFIG_ARM_MPU=n` to prj.conf.

I was able to compile & flash the application, and compile the hello_world extension.

Loading seems to be fine:
```
uart:~$ llext load_hex hello_world 7f454c460101010000000000000000000100280001000
0000000000000000000a80200000000000534000000000028000d000c0010b50348034ca0472a210
348a04710bd00000000000000001000000068656c6c6f20776f726c640a0000000041206e756d626
5722069732025750a00746573745f656e74727900002000000000000000004743433a20285a65706
879722053444b20302e31362e352d31292031322e322e3000412f000000616561626900012500000
00537452d4d00060d074d09021204140115011703180119011a011e0122012601000000000000000
00000000000000000000100000000000000000000000400f1ff00000000000000000000000003000
10000000000000000000000000003000300000000000000000000000000030004000000000000000
00000000000030005001300000000000000000000000000050016000000000000000000000000000
10013000000100000000000000000000100000000000000000000000000030006001300000000000
00000000000000006001900000000000000080000000100060000000000000000000000000003000
8000000000000000000000000000300090028000000010000001c000000120001003300000000000
00000000000100000000068656c6c6f5f776f726c645f6578742e6300246400247400746573745f6
56e7472795f73796d00746573745f656e747279007072696e746b000000100000000205000014000
000020f00001800000002050000000000000205000004000000020e0000002e73796d746162002e7
37472746162002e7368737472746162002e72656c2e74657874002e64617461002e627373002e726
f646174612e737472312e34002e72656c2e6578706f727465645f73796d002e636f6d6d656e74002
e41524d2e61747472696275746573000000000000000000000000000000000000000000000000000
000000000000000000000000000000000001f000000010000000600000000000000340000001c000
000000000000000000004000000000000001b0000000900000040000000000000001402000018000
0000a000000010000000400000008000000250000000100000003000000000000005000000000000
000000000000000000001000000000000002b0000000800000003000000000000005000000000000
0000000000000000000010000000000000030000000010000003200000000000000500000002b000
00000000000000000000400000001000000430000000100000002000000000000007c00000008000
000000000000000000004000000000000003f0000000900000040000000000000002c02000010000
0000a000000060000000400000008000000510000000100000030000000000000008400000023000
000000000000000000001000000010000005a000000030000700000000000000000a700000030000
0000000000000000000010000000000000001000000020000000000000000000000d800000000010
0000b0000000e000000040000001000000009000000030000000000000000000000d80100003a000
00000000000000000000100000000000000110000000300000000000000000000003c0200006a000
00000000000000000000100000000000000
[00:00:10.202,117] <dbg> llext_shell: cmd_llext_load_hex: hex2bin hex len 2400, llext buf sz 8192, read 1200
[00:00:10.212,341] <dbg> llext_shell: cmd_llext_load_hex: 4 byte MAGIC
                                      7f 45 4c 46                                      |.ELF             
[00:00:10.228,668] <dbg> llext: llext_load: Loading relocatable or shared elf
[00:00:10.236,297] <dbg> llext: do_llext_load: Finding ELF tables...
[00:00:10.243,103] <dbg> llext: llext_find_tables: section 0 at 2a8: name 0, type 0, flags 0, addr 0, size 0
[00:00:10.253,387] <dbg> llext: llext_find_tables: section 1 at 2d0: name 31, type 1, flags 6, addr 0, size 28
[00:00:10.263,854] <dbg> llext: llext_find_tables: section 2 at 2f8: name 27, type 9, flags 40, addr 0, size 24
[00:00:10.274,383] <dbg> llext: llext_find_tables: section 3 at 320: name 37, type 1, flags 3, addr 0, size 0
[00:00:10.284,759] <dbg> llext: llext_find_tables: section 4 at 348: name 43, type 8, flags 3, addr 0, size 0
[00:00:10.295,135] <dbg> llext: llext_find_tables: section 5 at 370: name 48, type 1, flags 32, addr 0, size 43
[00:00:10.305,694] <dbg> llext: llext_find_tables: section 6 at 398: name 67, type 1, flags 2, addr 0, size 8
[00:00:10.316,070] <dbg> llext: llext_find_tables: section 7 at 3c0: name 63, type 9, flags 40, addr 0, size 16
[00:00:10.326,599] <dbg> llext: llext_find_tables: section 8 at 3e8: name 81, type 1, flags 30, addr 0, size 35
[00:00:10.337,158] <dbg> llext: llext_find_tables: section 9 at 410: name 90, type 1879048195, flags 0, addr 0, size 48
[00:00:10.348,388] <dbg> llext: llext_find_tables: section 10 at 438: name 1, type 2, flags 0, addr 0, size 256
[00:00:10.358,917] <dbg> llext: llext_find_tables: symtab at 10
[00:00:10.365,325] <dbg> llext: llext_find_tables: section 11 at 460: name 9, type 3, flags 0, addr 0, size 58
[00:00:10.375,732] <dbg> llext: llext_find_tables: strtab at 11
[00:00:10.382,141] <dbg> llext: llext_find_tables: section 12 at 488: name 17, type 3, flags 0, addr 0, size 106
[00:00:10.392,730] <dbg> llext: llext_find_tables: shstrtab at 12
[00:00:10.399,261] <dbg> llext: do_llext_load: Allocate and copy strings...
[00:00:10.406,707] <dbg> llext: do_llext_load: Mapping ELF sections...
[00:00:10.413,665] <dbg> llext: llext_map_sections: section 0 name 
[00:00:10.420,379] <dbg> llext: llext_map_sections: Not copied section 
[00:00:10.427,429] <dbg> llext: llext_map_sections: section 1 name .text
[00:00:10.434,570] <dbg> llext: llext_map_sections: section 2 name .rel.text
[00:00:10.442,077] <dbg> llext: llext_map_sections: Not copied section .rel.text
[00:00:10.449,920] <dbg> llext: llext_map_sections: section 3 name .data
[00:00:10.457,061] <dbg> llext: llext_map_sections: section 4 name .bss
[00:00:10.464,141] <dbg> llext: llext_map_sections: section 5 name .rodata.str1.4
[00:00:10.472,045] <dbg> llext: llext_map_sections: Not copied section .rodata.str1.4
[00:00:10.480,346] <dbg> llext: llext_map_sections: section 6 name .exported_sym
[00:00:10.488,189] <dbg> llext: llext_map_sections: section 7 name .rel.exported_sym
[00:00:10.496,368] <dbg> llext: llext_map_sections: Not copied section .rel.exported_sym
[00:00:10.504,913] <dbg> llext: llext_map_sections: section 8 name .comment
[00:00:10.512,329] <dbg> llext: llext_map_sections: Not copied section .comment
[00:00:10.520,080] <dbg> llext: llext_map_sections: section 9 name .ARM.attributes
[00:00:10.528,106] <dbg> llext: llext_map_sections: Not copied section .ARM.attributes
[00:00:10.536,468] <dbg> llext: llext_map_sections: section 10 name .symtab
[00:00:10.543,853] <dbg> llext: llext_map_sections: Not copied section .symtab
[00:00:10.551,544] <dbg> llext: llext_map_sections: section 11 name .strtab
[00:00:10.558,929] <dbg> llext: llext_map_sections: Not copied section .strtab
[00:00:10.566,619] <dbg> llext: llext_map_sections: section 12 name .shstrtab
[00:00:10.574,188] <dbg> llext: llext_map_sections: Not copied section .shstrtab
[00:00:10.582,031] <dbg> llext: do_llext_load: Allocate and copy sections...
[00:00:10.589,569] <dbg> llext: do_llext_load: Counting exported symbols...
[00:00:10.596,923] <dbg> llext: llext_count_export_syms: symbol count 16
[00:00:10.604,125] <dbg> llext: llext_count_export_syms: unhandled symbol 1, name hello_world_ext.c, type tag 4, bind 0, sect 65521
[00:00:10.616,394] <dbg> llext: llext_count_export_syms: unhandled symbol 2, name , type tag 3, bind 0, sect 1
[00:00:10.626,861] <dbg> llext: llext_count_export_syms: unhandled symbol 3, name , type tag 3, bind 0, sect 3
[00:00:10.637,298] <dbg> llext: llext_count_export_syms: unhandled symbol 4, name , type tag 3, bind 0, sect 4
[00:00:10.647,766] <dbg> llext: llext_count_export_syms: unhandled symbol 5, name , type tag 3, bind 0, sect 5
[00:00:10.658,203] <dbg> llext: llext_count_export_syms: unhandled symbol 6, name $d, type tag 0, bind 0, sect 5
[00:00:10.668,853] <dbg> llext: llext_count_export_syms: unhandled symbol 7, name $t, type tag 0, bind 0, sect 1
[00:00:10.679,473] <dbg> llext: llext_count_export_syms: unhandled symbol 8, name $d, type tag 0, bind 0, sect 1
[00:00:10.690,093] <dbg> llext: llext_count_export_syms: unhandled symbol 9, name , type tag 3, bind 0, sect 6
[00:00:10.700,561] <dbg> llext: llext_count_export_syms: unhandled symbol 10, name $d, type tag 0, bind 0, sect 6
[00:00:10.711,273] <dbg> llext: llext_count_export_syms: unhandled symbol 11, name test_entry_sym, type tag 1, bind 0, sect 6
[00:00:10.723,022] <dbg> llext: llext_count_export_syms: unhandled symbol 12, name , type tag 3, bind 0, sect 8
[00:00:10.733,551] <dbg> llext: llext_count_export_syms: unhandled symbol 13, name , type tag 3, bind 0, sect 9
[00:00:10.744,110] <dbg> llext: llext_count_export_syms: function symbol 14, name test_entry, type tag 2, bind 1, sect 1
[00:00:10.755,432] <dbg> llext: llext_count_export_syms: unhandled symbol 15, name printk, type tag 0, bind 1, sect 0
[00:00:10.766,448] <dbg> llext: do_llext_load: Allocating memory for symbol table...
[00:00:10.774,658] <dbg> llext: do_llext_load: Copying symbols...
[00:00:10.781,250] <dbg> llext: llext_copy_symbols: function symbol 0 name test_entry addr 0x20008821
[00:00:10.790,863] <dbg> llext: do_llext_load: Linking ELF...
[00:00:10.797,088] <dbg> llext: llext_link: relocation section .rel.text (2) linked to section 10 has 3 relocations
[00:00:10.808,013] <dbg> llext: llext_link: relocation 2:0 info 502 (type 2, sym 5) offset 16 sym_name  sym_type 3 sym_bind 0 sym_ndx 5
[00:00:10.820,617] <inf> llext: found section symbol  addr 0x20008820
[00:00:10.827,514] <inf> llext: writing relocation symbol  type 2 sym 5 at addr 0x20008830 addr 0x20008820
[00:00:10.837,585] <dbg> elf: arch_elf_relocate: 2 20008830 20008820 
[00:00:10.844,512] <dbg> llext: llext_link: relocation 2:1 info f02 (type 2, sym 15) offset 20 sym_name printk sym_type 0 sym_bind 1 sym_ndx 0
[00:00:10.857,727] <inf> llext: found symbol printk at 0xbeb7
[00:00:10.863,922] <inf> llext: writing relocation symbol printk type 2 sym 15 at addr 0x20008834 addr 0xbeb7
[00:00:10.874,267] <dbg> elf: arch_elf_relocate: 2 20008834 beb7 printk
[00:00:10.881,378] <dbg> llext: llext_link: relocation 2:2 info 502 (type 2, sym 5) offset 24 sym_name  sym_type 3 sym_bind 0 sym_ndx 5
[00:00:10.893,981] <inf> llext: found section symbol  addr 0x20008820
[00:00:10.900,878] <inf> llext: writing relocation symbol  type 2 sym 5 at addr 0x20008838 addr 0x20008820
[00:00:10.910,980] <dbg> elf: arch_elf_relocate: 2 20008838 20008820 
[00:00:10.917,907] <dbg> llext: llext_link: relocation section .rel.exported_sym (7) linked to section 10 has 2 relocations
[00:00:10.929,504] <dbg> llext: llext_link: relocation 7:0 info 502 (type 2, sym 5) offset 0 sym_name  sym_type 3 sym_bind 0 sym_ndx 5
[00:00:10.942,016] <inf> llext: found section symbol  addr 0x20008820
[00:00:10.948,913] <inf> llext: writing relocation symbol  type 2 sym 5 at addr 0x20008860 addr 0x20008820
[00:00:10.959,014] <dbg> elf: arch_elf_relocate: 2 20008860 20008820 
[00:00:10.965,942] <dbg> llext: llext_link: relocation 7:1 info e02 (type 2, sym 14) offset 4 sym_name test_entry sym_type 2 sym_bind 1 sym_ndx 1
[00:00:10.979,431] <inf> llext: found section symbol test_entry addr 0x20008821
[00:00:10.987,182] <inf> llext: writing relocation symbol test_entry type 2 sym 14 at addr 0x20008864 addr 0x20008821
[00:00:10.998,229] <dbg> elf: arch_elf_relocate: 2 20008864 20008821 test_entry
[00:00:11.006,011] <dbg> llext: llext_export_symbols: sym 0x20008821 name  in 0x200087a4
[00:00:11.014,617] <dbg> llext: do_llext_load: loaded module, .text at 0x20008820, .rodata at 0
[00:00:11.023,773] <inf> llext: Loaded extension hello_world
Successfully loaded extension hello_world, addr 0x2000867c
```

The module is listed:
```
uart:~$ llext list
| Name             | Size         |
|      hello_world |          533 |
```

However, the symbol is not listed:
```
uart:~$ llext list_symbols hello_world
Extension: hello_world symbols
| Symbol           | Address    |
uart:~$ 
```

Calling the function shows nothing.

markus-becker-tridonic-com commented:
Same result if I add `LL_EXTENSION_SYMBOL(hello_world);` at the end.

teburd commented:
> @teburd I was trying this on an nRF52840DK (using Nordic Connect SDK and the bundled Zephyr) and cannot get it to work.
> 
> I have added `CONFIG_ARM_MPU=n` to prj.conf.
> 
> I was able to compile & flash the application, and compile the hello_world extension.
> 
> Loading seems to be fine:
> 
> ```
> uart:~$ llext load_hex hello_world 7f454c460101010000000000000000000100280001000
> 0000000000000000000a80200000000000534000000000028000d000c0010b50348034ca0472a210
> 348a04710bd00000000000000001000000068656c6c6f20776f726c640a0000000041206e756d626
> 5722069732025750a00746573745f656e74727900002000000000000000004743433a20285a65706
> 879722053444b20302e31362e352d31292031322e322e3000412f000000616561626900012500000
> 00537452d4d00060d074d09021204140115011703180119011a011e0122012601000000000000000
> 00000000000000000000100000000000000000000000400f1ff00000000000000000000000003000
> 10000000000000000000000000003000300000000000000000000000000030004000000000000000
> 00000000000030005001300000000000000000000000000050016000000000000000000000000000
> 10013000000100000000000000000000100000000000000000000000000030006001300000000000
> 00000000000000006001900000000000000080000000100060000000000000000000000000003000
> 8000000000000000000000000000300090028000000010000001c000000120001003300000000000
> 00000000000100000000068656c6c6f5f776f726c645f6578742e6300246400247400746573745f6
> 56e7472795f73796d00746573745f656e747279007072696e746b000000100000000205000014000
> 000020f00001800000002050000000000000205000004000000020e0000002e73796d746162002e7
> 37472746162002e7368737472746162002e72656c2e74657874002e64617461002e627373002e726
> f646174612e737472312e34002e72656c2e6578706f727465645f73796d002e636f6d6d656e74002
> e41524d2e61747472696275746573000000000000000000000000000000000000000000000000000
> 000000000000000000000000000000000001f000000010000000600000000000000340000001c000
> 000000000000000000004000000000000001b0000000900000040000000000000001402000018000
> 0000a000000010000000400000008000000250000000100000003000000000000005000000000000
> 000000000000000000001000000000000002b0000000800000003000000000000005000000000000
> 0000000000000000000010000000000000030000000010000003200000000000000500000002b000
> 00000000000000000000400000001000000430000000100000002000000000000007c00000008000
> 000000000000000000004000000000000003f0000000900000040000000000000002c02000010000
> 0000a000000060000000400000008000000510000000100000030000000000000008400000023000
> 000000000000000000001000000010000005a000000030000700000000000000000a700000030000
> 0000000000000000000010000000000000001000000020000000000000000000000d800000000010
> 0000b0000000e000000040000001000000009000000030000000000000000000000d80100003a000
> 00000000000000000000100000000000000110000000300000000000000000000003c0200006a000
> 00000000000000000000100000000000000
> [00:00:10.202,117] <dbg> llext_shell: cmd_llext_load_hex: hex2bin hex len 2400, llext buf sz 8192, read 1200
> [00:00:10.212,341] <dbg> llext_shell: cmd_llext_load_hex: 4 byte MAGIC
>                                       7f 45 4c 46                                      |.ELF             
> [00:00:10.228,668] <dbg> llext: llext_load: Loading relocatable or shared elf
> [00:00:10.236,297] <dbg> llext: do_llext_load: Finding ELF tables...
> [00:00:10.243,103] <dbg> llext: llext_find_tables: section 0 at 2a8: name 0, type 0, flags 0, addr 0, size 0
> [00:00:10.253,387] <dbg> llext: llext_find_tables: section 1 at 2d0: name 31, type 1, flags 6, addr 0, size 28
> [00:00:10.263,854] <dbg> llext: llext_find_tables: section 2 at 2f8: name 27, type 9, flags 40, addr 0, size 24
> [00:00:10.274,383] <dbg> llext: llext_find_tables: section 3 at 320: name 37, type 1, flags 3, addr 0, size 0
> [00:00:10.284,759] <dbg> llext: llext_find_tables: section 4 at 348: name 43, type 8, flags 3, addr 0, size 0
> [00:00:10.295,135] <dbg> llext: llext_find_tables: section 5 at 370: name 48, type 1, flags 32, addr 0, size 43
> [00:00:10.305,694] <dbg> llext: llext_find_tables: section 6 at 398: name 67, type 1, flags 2, addr 0, size 8
> [00:00:10.316,070] <dbg> llext: llext_find_tables: section 7 at 3c0: name 63, type 9, flags 40, addr 0, size 16
> [00:00:10.326,599] <dbg> llext: llext_find_tables: section 8 at 3e8: name 81, type 1, flags 30, addr 0, size 35
> [00:00:10.337,158] <dbg> llext: llext_find_tables: section 9 at 410: name 90, type 1879048195, flags 0, addr 0, size 48
> [00:00:10.348,388] <dbg> llext: llext_find_tables: section 10 at 438: name 1, type 2, flags 0, addr 0, size 256
> [00:00:10.358,917] <dbg> llext: llext_find_tables: symtab at 10
> [00:00:10.365,325] <dbg> llext: llext_find_tables: section 11 at 460: name 9, type 3, flags 0, addr 0, size 58
> [00:00:10.375,732] <dbg> llext: llext_find_tables: strtab at 11
> [00:00:10.382,141] <dbg> llext: llext_find_tables: section 12 at 488: name 17, type 3, flags 0, addr 0, size 106
> [00:00:10.392,730] <dbg> llext: llext_find_tables: shstrtab at 12
> [00:00:10.399,261] <dbg> llext: do_llext_load: Allocate and copy strings...
> [00:00:10.406,707] <dbg> llext: do_llext_load: Mapping ELF sections...
> [00:00:10.413,665] <dbg> llext: llext_map_sections: section 0 name 
> [00:00:10.420,379] <dbg> llext: llext_map_sections: Not copied section 
> [00:00:10.427,429] <dbg> llext: llext_map_sections: section 1 name .text
> [00:00:10.434,570] <dbg> llext: llext_map_sections: section 2 name .rel.text
> [00:00:10.442,077] <dbg> llext: llext_map_sections: Not copied section .rel.text
> [00:00:10.449,920] <dbg> llext: llext_map_sections: section 3 name .data
> [00:00:10.457,061] <dbg> llext: llext_map_sections: section 4 name .bss
> [00:00:10.464,141] <dbg> llext: llext_map_sections: section 5 name .rodata.str1.4
> [00:00:10.472,045] <dbg> llext: llext_map_sections: Not copied section .rodata.str1.4
> [00:00:10.480,346] <dbg> llext: llext_map_sections: section 6 name .exported_sym
> [00:00:10.488,189] <dbg> llext: llext_map_sections: section 7 name .rel.exported_sym
> [00:00:10.496,368] <dbg> llext: llext_map_sections: Not copied section .rel.exported_sym
> [00:00:10.504,913] <dbg> llext: llext_map_sections: section 8 name .comment
> [00:00:10.512,329] <dbg> llext: llext_map_sections: Not copied section .comment
> [00:00:10.520,080] <dbg> llext: llext_map_sections: section 9 name .ARM.attributes
> [00:00:10.528,106] <dbg> llext: llext_map_sections: Not copied section .ARM.attributes
> [00:00:10.536,468] <dbg> llext: llext_map_sections: section 10 name .symtab
> [00:00:10.543,853] <dbg> llext: llext_map_sections: Not copied section .symtab
> [00:00:10.551,544] <dbg> llext: llext_map_sections: section 11 name .strtab
> [00:00:10.558,929] <dbg> llext: llext_map_sections: Not copied section .strtab
> [00:00:10.566,619] <dbg> llext: llext_map_sections: section 12 name .shstrtab
> [00:00:10.574,188] <dbg> llext: llext_map_sections: Not copied section .shstrtab
> [00:00:10.582,031] <dbg> llext: do_llext_load: Allocate and copy sections...
> [00:00:10.589,569] <dbg> llext: do_llext_load: Counting exported symbols...
> [00:00:10.596,923] <dbg> llext: llext_count_export_syms: symbol count 16
> [00:00:10.604,125] <dbg> llext: llext_count_export_syms: unhandled symbol 1, name hello_world_ext.c, type tag 4, bind 0, sect 65521
> [00:00:10.616,394] <dbg> llext: llext_count_export_syms: unhandled symbol 2, name , type tag 3, bind 0, sect 1
> [00:00:10.626,861] <dbg> llext: llext_count_export_syms: unhandled symbol 3, name , type tag 3, bind 0, sect 3
> [00:00:10.637,298] <dbg> llext: llext_count_export_syms: unhandled symbol 4, name , type tag 3, bind 0, sect 4
> [00:00:10.647,766] <dbg> llext: llext_count_export_syms: unhandled symbol 5, name , type tag 3, bind 0, sect 5
> [00:00:10.658,203] <dbg> llext: llext_count_export_syms: unhandled symbol 6, name $d, type tag 0, bind 0, sect 5
> [00:00:10.668,853] <dbg> llext: llext_count_export_syms: unhandled symbol 7, name $t, type tag 0, bind 0, sect 1
> [00:00:10.679,473] <dbg> llext: llext_count_export_syms: unhandled symbol 8, name $d, type tag 0, bind 0, sect 1
> [00:00:10.690,093] <dbg> llext: llext_count_export_syms: unhandled symbol 9, name , type tag 3, bind 0, sect 6
> [00:00:10.700,561] <dbg> llext: llext_count_export_syms: unhandled symbol 10, name $d, type tag 0, bind 0, sect 6
> [00:00:10.711,273] <dbg> llext: llext_count_export_syms: unhandled symbol 11, name test_entry_sym, type tag 1, bind 0, sect 6
> [00:00:10.723,022] <dbg> llext: llext_count_export_syms: unhandled symbol 12, name , type tag 3, bind 0, sect 8
> [00:00:10.733,551] <dbg> llext: llext_count_export_syms: unhandled symbol 13, name , type tag 3, bind 0, sect 9
> [00:00:10.744,110] <dbg> llext: llext_count_export_syms: function symbol 14, name test_entry, type tag 2, bind 1, sect 1
> [00:00:10.755,432] <dbg> llext: llext_count_export_syms: unhandled symbol 15, name printk, type tag 0, bind 1, sect 0
> [00:00:10.766,448] <dbg> llext: do_llext_load: Allocating memory for symbol table...
> [00:00:10.774,658] <dbg> llext: do_llext_load: Copying symbols...
> [00:00:10.781,250] <dbg> llext: llext_copy_symbols: function symbol 0 name test_entry addr 0x20008821
> [00:00:10.790,863] <dbg> llext: do_llext_load: Linking ELF...
> [00:00:10.797,088] <dbg> llext: llext_link: relocation section .rel.text (2) linked to section 10 has 3 relocations
> [00:00:10.808,013] <dbg> llext: llext_link: relocation 2:0 info 502 (type 2, sym 5) offset 16 sym_name  sym_type 3 sym_bind 0 sym_ndx 5
> [00:00:10.820,617] <inf> llext: found section symbol  addr 0x20008820
> [00:00:10.827,514] <inf> llext: writing relocation symbol  type 2 sym 5 at addr 0x20008830 addr 0x20008820
> [00:00:10.837,585] <dbg> elf: arch_elf_relocate: 2 20008830 20008820 
> [00:00:10.844,512] <dbg> llext: llext_link: relocation 2:1 info f02 (type 2, sym 15) offset 20 sym_name printk sym_type 0 sym_bind 1 sym_ndx 0
> [00:00:10.857,727] <inf> llext: found symbol printk at 0xbeb7
> [00:00:10.863,922] <inf> llext: writing relocation symbol printk type 2 sym 15 at addr 0x20008834 addr 0xbeb7
> [00:00:10.874,267] <dbg> elf: arch_elf_relocate: 2 20008834 beb7 printk
> [00:00:10.881,378] <dbg> llext: llext_link: relocation 2:2 info 502 (type 2, sym 5) offset 24 sym_name  sym_type 3 sym_bind 0 sym_ndx 5
> [00:00:10.893,981] <inf> llext: found section symbol  addr 0x20008820
> [00:00:10.900,878] <inf> llext: writing relocation symbol  type 2 sym 5 at addr 0x20008838 addr 0x20008820
> [00:00:10.910,980] <dbg> elf: arch_elf_relocate: 2 20008838 20008820 
> [00:00:10.917,907] <dbg> llext: llext_link: relocation section .rel.exported_sym (7) linked to section 10 has 2 relocations
> [00:00:10.929,504] <dbg> llext: llext_link: relocation 7:0 info 502 (type 2, sym 5) offset 0 sym_name  sym_type 3 sym_bind 0 sym_ndx 5
> [00:00:10.942,016] <inf> llext: found section symbol  addr 0x20008820
> [00:00:10.948,913] <inf> llext: writing relocation symbol  type 2 sym 5 at addr 0x20008860 addr 0x20008820
> [00:00:10.959,014] <dbg> elf: arch_elf_relocate: 2 20008860 20008820 
> [00:00:10.965,942] <dbg> llext: llext_link: relocation 7:1 info e02 (type 2, sym 14) offset 4 sym_name test_entry sym_type 2 sym_bind 1 sym_ndx 1
> [00:00:10.979,431] <inf> llext: found section symbol test_entry addr 0x20008821
> [00:00:10.987,182] <inf> llext: writing relocation symbol test_entry type 2 sym 14 at addr 0x20008864 addr 0x20008821
> [00:00:10.998,229] <dbg> elf: arch_elf_relocate: 2 20008864 20008821 test_entry
> [00:00:11.006,011] <dbg> llext: llext_export_symbols: sym 0x20008821 name  in 0x200087a4
> [00:00:11.014,617] <dbg> llext: do_llext_load: loaded module, .text at 0x20008820, .rodata at 0
> [00:00:11.023,773] <inf> llext: Loaded extension hello_world
> Successfully loaded extension hello_world, addr 0x2000867c
> ```
> 
> The module is listed:
> 
> ```
> uart:~$ llext list
> | Name             | Size         |
> |      hello_world |          533 |
> ```
> 
> However, the symbol is not listed:
> 
> ```
> uart:~$ llext list_symbols hello_world
> Extension: hello_world symbols
> | Symbol           | Address    |
> uart:~$ 
> ```
> 
> Calling the function shows nothing.

Yes this is the issue, building the extension currently requires all of zephyr which isn't quite right. It didn't need to be that way before but since we've added more specific "export" symbols rather than simply using the ELF global symbol table we now need some linker script and macro stuff to make the sample work which is... I have to say kind of moving away from the simple sample this was meant to be.

I haven't come up with a great solution to this other than perhaps building the extension as part of the sample as we do with the tests... but this just feels like its going to mask whats really going on which I was really happy to expose and explain in the sample.

teburd commented:
I've updated the PR to use west+ninja (cmake with add_llext_target) to show how to build extensions now rather than directly calling out to gcc. The looking glass has become a bit translucent rather than transparent in doing so, but this is how we do build extensions now and likely will continue to be done this way well into the future.

pillo79 commented:
Thanks for the fix. This could be backported to LTS as well I guess?

markus-becker-tridonic-com commented on samples/subsys/llext/shell_loader/README.rst:
```diff
@@ -49,18 +49,21 @@ which can be seen with llext help
-  $ arm-zephyr-eabi-gcc -mlong-calls -mthumb -c -o hello_world.elf tests/subsys/llext/hello_world/hello_world.c
+  $ west build -p always -b mimxrt1010_evk samples/subsys/llext/shell_loader
+  $ cd build && ninja -vvv hello_world_ext
   $ arm-zephyr-eabi-objdump -r -d -x hello_world.elf
```
Isn't the resulting file called hello_world_ext.llext.pkg_input now?

markus-becker-tridonic-com replied:
Also below for the xxd command, it would need to be changed.

pillo79 commented:
> building the extension currently requires all of zephyr which isn't quite right. It didn't need to be that way before but since we've added more specific "export" symbols rather than simply using the ELF global symbol table we now need some linker script and macro stuff to make the sample work which is... I have to say kind of moving away from the simple sample this was meant to be.

I played with this a bit to try to improve the situation and came up with an alternative solution at #84393. PTAL and let me know your feedback there!

teburd commented:
Marked DNM as I quite like the changes @pillo79 has done to get back to the original functionality this was using

kartben commented:
superseded by https://github.com/zephyrproject-rtos/zephyr/pull/84393
